test/Chainweb/Test/Cut/TestBlockDb.hs

```diff
@@ -64,13 +63,34 @@ mkTestBlockDb cv rdb = do
     return $! TestBlockDb wdb pdb initCut
 
 -- | Add a block.
-addTestBlockDb :: TestBlockDb -> Nonce -> GenBlockTime -> ChainId -> PayloadWithOutputs -> IO ()
+--
+-- Returns False when mining fails due to BadAdjacents, which usually means that
+-- the chain is blocked. Retry with another chain!
+--
+addTestBlockDb
+    :: TestBlockDb
+    -> Nonce
+    -> GenBlockTime
+    -> ChainId
+    -> PayloadWithOutputs
+    -> IO Bool
 addTestBlockDb (TestBlockDb wdb pdb cmv) n gbt cid outs = do
   c <- takeMVar cmv
   r <- testMine' wdb n gbt (_payloadWithOutputsPayloadHash outs) cid c
-  (T2 _ c') <- fromEitherM $ first (userError . show) $ r
-  casInsert pdb outs
-  putMVar cmv c'
+  case r of
+    -- success
+    Right (T2 _ c') -> do
+        casInsert pdb outs
+        putMVar cmv c'
+        return True
+
+    -- mining failed, probably because chain is blocked
+    Left BadAdjacents -> do
+        putMVar cmv c
+        return False
```

Comment:
Before the function would fail with an `userError` in this case. This change informs the caller when mining fails.

In practice there are tests where it can happen that the mining of a blocked chain is attempted. It was by pure chance that this didn't showed up before. The latest version of `hashable` changed the hash value of chain ids and thus the ordering of chains in tests, which caused this issue to surface.

Reply:
What do you think about applying a "if comments on github make it easy to understand the PR, they should be in the code" rule here?

Reply:
I am not sure in case of this comments, since they talk an old version of the code and the diff with the new version. I think, PR comments my be a better place for those.

But, generally, I agree, that comments that explain the behavior of the new code should be code comments.